pkg/liqo-controller-manager/virtualNode-controller/namespaceMap_lifecycle.go

```diff
@@ -0,0 +1,128 @@
+	ctrlutils.RemoveFinalizer(&nm, namespaceMapFinalizer)
+	klog.Infof("The NamespaceMap '%s' is requested to be deleted", nm.GetName())
+
+	if err := r.Update(context.TODO(), &nm); err != nil {
+		// WARNING: Is possible that this Update is called on a resource that is no more here
+		// it doesn't return "NotFound" but "Conflict"
+		klog.Errorf(" %s --> Problems while removing finalizer from '%s'", err, nm.GetName())
+		return err
+	}
+	klog.Infof("Finalizer is correctly removed from the NamespaceMap '%s'", nm.GetName())
+
+	return nil
+}
+
+// create a new NamespaceMap with Finalizer and OwnerReference
+func (r *VirtualNodeReconciler) createNamespaceMap(n corev1.Node, s mapsv1alpha1.NamespaceMapStatus) error {
+	nm := &mapsv1alpha1.NamespaceMap{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "namespaceresources.liqo.io/v1",
+			Kind:       "NamespaceMap",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: fmt.Sprintf("%s-", n.GetAnnotations()[constctrl.VirtualNodeClusterId]),
+			Namespace:    constctrl.MapNamespaceName,
+			Labels: map[string]string{
+				constctrl.VirtualNodeClusterId: n.GetAnnotations()[constctrl.VirtualNodeClusterId],
+			},
+		},
+		Status: s,
+	}
+
+	ctrlutils.AddFinalizer(nm, namespaceMapFinalizer)
+	if err := ctrlutils.SetControllerReference(&n, nm, r.Scheme); err != nil {
+		return err
+	}
+
+	if err := r.Create(context.TODO(), nm); err != nil {
+		klog.Errorf("%s --> Problems in NamespaceMap creation for the virtual node '%s'", err, n.GetName())
+		return err
+	}
+	klog.Infof(" Create NamespaceMap '%s' for the virtual node '%s'", nm.GetName(), n.GetName())
+	return nil
+}
+
+// first create a new NamespaceMap which preserves the Status and then delete the other
+func (r *VirtualNodeReconciler) regenerateNamespaceMap(nm mapsv1alpha1.NamespaceMap, n corev1.Node) error {
+
+	// create a new namespaceMap with same Status but with different Name
+	if err := r.createNamespaceMap(n, nm.Status); err != nil {
+		return err
+	}
+
+	if err := r.deleteNamespaceMap(nm); err != nil {
+		return err
+	}
+	return nil
+}
+
+// This function manages NamespaceMaps Lifecycle on the basis of NamespaceMaps' number
+func (r *VirtualNodeReconciler) namespaceMapLifecycle(n corev1.Node) error {
+
+	nms := &mapsv1alpha1.NamespaceMapList{}
+	if err := r.List(context.TODO(), nms, client.InNamespace(constctrl.MapNamespaceName),
+		client.MatchingLabels{constctrl.VirtualNodeClusterId: n.GetAnnotations()[constctrl.VirtualNodeClusterId]}); err != nil {
+		klog.Errorf("%s --> Unable to List NamespaceMaps of virtual node '%s'", err, n.GetName())
+		return err
+	}
+
+	if len(nms.Items) == 0 {
+		return r.createNamespaceMap(n, mapsv1alpha1.NamespaceMapStatus{})
+	}
+
+	if len(nms.Items) == 1 {
+		if !nms.Items[0].GetDeletionTimestamp().IsZero() {
+			return r.regenerateNamespaceMap(nms.Items[0], n)
+		}
+		return nil
+	}
+
+	if len(nms.Items) > 1 {
+
+		oldestCreation := metav1.Time{Time: time.Now()}
+		oldestMap := -1
```

Comment:
Why do you need to initialize oldestMap to -1? It seems this initialization is ineffective but badly affects the code clarity.

Reply:
yes you are right, thank you